{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook for analyzing datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene Expression Dataset 1 (GEX)\n",
    "gex1 = sc.read_h5ad('../multiome/multiome_gex_processed_training.h5ad')\n",
    "\n",
    "# DNA Accessibility Dataset (ATAC)\n",
    "atac = sc.read_h5ad('../multiome/multiome_atac_processed_training.h5ad')\n",
    "\n",
    "\n",
    "# Gene Expression Dataset 2 (GEX)\n",
    "gex2 = sc.read_h5ad('../cite/cite_gex_processed_training.h5ad')\n",
    "\n",
    "# Protein Abundance Dataset (ADT)\n",
    "adt = sc.read_h5ad('../cite/cite_adt_processed_training.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene expression data (GEX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22463 × 13431\n",
       "    obs: 'n_genes_by_counts', 'pct_counts_mt', 'n_counts', 'n_genes', 'size_factors', 'phase', 'cell_type', 'pseudotime_order_GEX', 'batch'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'\n",
       "    uns: 'dataset_id', 'organism'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    layers: 'counts', 'log_norm'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gex1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22463, 13431)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Data Matrix, each cell is a row and each column is a feature (in this case the expression value of a gene) \n",
    "\n",
    "gex1.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>size_factors</th>\n",
       "      <th>phase</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>pseudotime_order_GEX</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAGTTGTCACCCTCAC-1-s1d1</th>\n",
       "      <td>1022</td>\n",
       "      <td>1.061008</td>\n",
       "      <td>1508.0</td>\n",
       "      <td>1022</td>\n",
       "      <td>0.453484</td>\n",
       "      <td>S</td>\n",
       "      <td>Naive CD20+ B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTATGGCCATAACGGG-1-s1d1</th>\n",
       "      <td>1081</td>\n",
       "      <td>0.604230</td>\n",
       "      <td>1655.0</td>\n",
       "      <td>1081</td>\n",
       "      <td>0.455631</td>\n",
       "      <td>G2M</td>\n",
       "      <td>CD14+ Mono</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGCACACAGGTTAAA-1-s1d1</th>\n",
       "      <td>3304</td>\n",
       "      <td>0.650069</td>\n",
       "      <td>7230.0</td>\n",
       "      <td>3304</td>\n",
       "      <td>2.435348</td>\n",
       "      <td>G2M</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATTTGGTAATGGAA-1-s1d1</th>\n",
       "      <td>793</td>\n",
       "      <td>0.812274</td>\n",
       "      <td>1108.0</td>\n",
       "      <td>793</td>\n",
       "      <td>0.347226</td>\n",
       "      <td>G2M</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACATAGGTGTCCA-1-s1d1</th>\n",
       "      <td>1219</td>\n",
       "      <td>1.674770</td>\n",
       "      <td>1851.0</td>\n",
       "      <td>1219</td>\n",
       "      <td>0.534205</td>\n",
       "      <td>G2M</td>\n",
       "      <td>CD16+ Mono</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTAAGCAACTAGGG-8-s3d6</th>\n",
       "      <td>1313</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1937.0</td>\n",
       "      <td>1313</td>\n",
       "      <td>0.522115</td>\n",
       "      <td>S</td>\n",
       "      <td>HSC</td>\n",
       "      <td>0.098024</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTCCTTCGGCTAGC-8-s3d6</th>\n",
       "      <td>2846</td>\n",
       "      <td>0.015901</td>\n",
       "      <td>6289.0</td>\n",
       "      <td>2846</td>\n",
       "      <td>1.737451</td>\n",
       "      <td>G2M</td>\n",
       "      <td>CD4+ T activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTGCGTTGTTGGA-8-s3d6</th>\n",
       "      <td>983</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1390.0</td>\n",
       "      <td>983</td>\n",
       "      <td>0.291723</td>\n",
       "      <td>G2M</td>\n",
       "      <td>pDC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCCTCCCAGCCAGTT-8-s3d6</th>\n",
       "      <td>1351</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2190.0</td>\n",
       "      <td>1351</td>\n",
       "      <td>0.585303</td>\n",
       "      <td>G2M</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTGAACCATCCCGCT-8-s3d6</th>\n",
       "      <td>1661</td>\n",
       "      <td>0.068259</td>\n",
       "      <td>2930.0</td>\n",
       "      <td>1661</td>\n",
       "      <td>0.746938</td>\n",
       "      <td>G2M</td>\n",
       "      <td>CD4+ T activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22463 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         n_genes_by_counts  pct_counts_mt  n_counts  n_genes  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1               1022       1.061008    1508.0     1022   \n",
       "CTATGGCCATAACGGG-1-s1d1               1081       0.604230    1655.0     1081   \n",
       "CCGCACACAGGTTAAA-1-s1d1               3304       0.650069    7230.0     3304   \n",
       "TCATTTGGTAATGGAA-1-s1d1                793       0.812274    1108.0      793   \n",
       "ACCACATAGGTGTCCA-1-s1d1               1219       1.674770    1851.0     1219   \n",
       "...                                    ...            ...       ...      ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6               1313       0.000000    1937.0     1313   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6               2846       0.015901    6289.0     2846   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                983       0.000000    1390.0      983   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6               1351       0.000000    2190.0     1351   \n",
       "AGTGAACCATCCCGCT-8-s3d6               1661       0.068259    2930.0     1661   \n",
       "\n",
       "                         size_factors phase         cell_type  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1      0.453484     S     Naive CD20+ B   \n",
       "CTATGGCCATAACGGG-1-s1d1      0.455631   G2M        CD14+ Mono   \n",
       "CCGCACACAGGTTAAA-1-s1d1      2.435348   G2M            CD8+ T   \n",
       "TCATTTGGTAATGGAA-1-s1d1      0.347226   G2M            CD8+ T   \n",
       "ACCACATAGGTGTCCA-1-s1d1      0.534205   G2M        CD16+ Mono   \n",
       "...                               ...   ...               ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6      0.522115     S               HSC   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6      1.737451   G2M  CD4+ T activated   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6      0.291723   G2M               pDC   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6      0.585303   G2M            CD8+ T   \n",
       "AGTGAACCATCCCGCT-8-s3d6      0.746938   G2M  CD4+ T activated   \n",
       "\n",
       "                         pseudotime_order_GEX batch  \n",
       "TAGTTGTCACCCTCAC-1-s1d1                   NaN  s1d1  \n",
       "CTATGGCCATAACGGG-1-s1d1                   NaN  s1d1  \n",
       "CCGCACACAGGTTAAA-1-s1d1                   NaN  s1d1  \n",
       "TCATTTGGTAATGGAA-1-s1d1                   NaN  s1d1  \n",
       "ACCACATAGGTGTCCA-1-s1d1                   NaN  s1d1  \n",
       "...                                       ...   ...  \n",
       "TAGTAAGCAACTAGGG-8-s3d6              0.098024  s3d6  \n",
       "TGGTCCTTCGGCTAGC-8-s3d6                   NaN  s3d6  \n",
       "CGCTTGCGTTGTTGGA-8-s3d6                   NaN  s3d6  \n",
       "ACCCTCCCAGCCAGTT-8-s3d6                   NaN  s3d6  \n",
       "AGTGAACCATCCCGCT-8-s3d6                   NaN  s3d6  \n",
       "\n",
       "[22463 rows x 9 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This object also stores metadata for the cells as a dataframe\n",
    "\n",
    "gex1.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.5</th>\n",
       "      <td>ENSG00000241860</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01409</th>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01128</th>\n",
       "      <td>ENSG00000228794</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL17</th>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND5</th>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL592183.1</th>\n",
       "      <td>ENSG00000273748</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>GEX</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13431 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_ids feature_types  genome\n",
       "AL627309.5  ENSG00000241860           GEX  GRCh38\n",
       "LINC01409   ENSG00000237491           GEX  GRCh38\n",
       "LINC01128   ENSG00000228794           GEX  GRCh38\n",
       "NOC2L       ENSG00000188976           GEX  GRCh38\n",
       "KLHL17      ENSG00000187961           GEX  GRCh38\n",
       "...                     ...           ...     ...\n",
       "MT-ND5      ENSG00000198786           GEX  GRCh38\n",
       "MT-ND6      ENSG00000198695           GEX  GRCh38\n",
       "MT-CYB      ENSG00000198727           GEX  GRCh38\n",
       "AL592183.1  ENSG00000273748           GEX  GRCh38\n",
       "AC240274.1  ENSG00000271254           GEX  GRCh38\n",
       "\n",
       "[13431 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This object also stores metadata for the features(i.e. genes) as a dataframe\n",
    "\n",
    "gex1.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNA Accessibility data (ATAC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 22463 × 116490\n",
       "    obs: 'nCount_peaks', 'atac_fragments', 'reads_in_peaks_frac', 'blacklist_fraction', 'nucleosome_signal', 'cell_type', 'pseudotime_order_ATAC', 'batch'\n",
       "    var: 'feature_types'\n",
       "    uns: 'dataset_id', 'gene_activity_var_names', 'organism'\n",
       "    obsm: 'gene_activity', 'lsi_full', 'lsi_red', 'umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(22463, 116490)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Data Matrix, each cell is a row and each column is a feature (in this case the accessibility of the DNA at a specific position) \n",
    "\n",
    "atac.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nCount_peaks</th>\n",
       "      <th>atac_fragments</th>\n",
       "      <th>reads_in_peaks_frac</th>\n",
       "      <th>blacklist_fraction</th>\n",
       "      <th>nucleosome_signal</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>pseudotime_order_ATAC</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAGTTGTCACCCTCAC-1-s1d1</th>\n",
       "      <td>4031.0</td>\n",
       "      <td>5400</td>\n",
       "      <td>0.746481</td>\n",
       "      <td>0.003473</td>\n",
       "      <td>0.642468</td>\n",
       "      <td>Naive CD20+ B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTATGGCCATAACGGG-1-s1d1</th>\n",
       "      <td>8636.0</td>\n",
       "      <td>19266</td>\n",
       "      <td>0.448251</td>\n",
       "      <td>0.003126</td>\n",
       "      <td>1.220679</td>\n",
       "      <td>CD14+ Mono</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGCACACAGGTTAAA-1-s1d1</th>\n",
       "      <td>4674.0</td>\n",
       "      <td>6177</td>\n",
       "      <td>0.756678</td>\n",
       "      <td>0.001284</td>\n",
       "      <td>0.692573</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATTTGGTAATGGAA-1-s1d1</th>\n",
       "      <td>2803.0</td>\n",
       "      <td>4019</td>\n",
       "      <td>0.697437</td>\n",
       "      <td>0.000714</td>\n",
       "      <td>0.633838</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACATAGGTGTCCA-1-s1d1</th>\n",
       "      <td>1790.0</td>\n",
       "      <td>2568</td>\n",
       "      <td>0.697040</td>\n",
       "      <td>0.003352</td>\n",
       "      <td>0.727660</td>\n",
       "      <td>CD16+ Mono</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s1d1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTAAGCAACTAGGG-8-s3d6</th>\n",
       "      <td>7239.0</td>\n",
       "      <td>10580</td>\n",
       "      <td>0.684216</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0.866142</td>\n",
       "      <td>HSC</td>\n",
       "      <td>0.23367</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGTCCTTCGGCTAGC-8-s3d6</th>\n",
       "      <td>16056.0</td>\n",
       "      <td>22771</td>\n",
       "      <td>0.705107</td>\n",
       "      <td>0.000810</td>\n",
       "      <td>0.964006</td>\n",
       "      <td>CD4+ T activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTTGCGTTGTTGGA-8-s3d6</th>\n",
       "      <td>8388.0</td>\n",
       "      <td>14137</td>\n",
       "      <td>0.593337</td>\n",
       "      <td>0.000358</td>\n",
       "      <td>1.215227</td>\n",
       "      <td>pDC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCCTCCCAGCCAGTT-8-s3d6</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1381</td>\n",
       "      <td>0.724837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>CD8+ T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTGAACCATCCCGCT-8-s3d6</th>\n",
       "      <td>3183.0</td>\n",
       "      <td>4684</td>\n",
       "      <td>0.679547</td>\n",
       "      <td>0.001257</td>\n",
       "      <td>0.946565</td>\n",
       "      <td>CD4+ T activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>s3d6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22463 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         nCount_peaks  atac_fragments  reads_in_peaks_frac  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1        4031.0            5400             0.746481   \n",
       "CTATGGCCATAACGGG-1-s1d1        8636.0           19266             0.448251   \n",
       "CCGCACACAGGTTAAA-1-s1d1        4674.0            6177             0.756678   \n",
       "TCATTTGGTAATGGAA-1-s1d1        2803.0            4019             0.697437   \n",
       "ACCACATAGGTGTCCA-1-s1d1        1790.0            2568             0.697040   \n",
       "...                               ...             ...                  ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6        7239.0           10580             0.684216   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6       16056.0           22771             0.705107   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6        8388.0           14137             0.593337   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6        1001.0            1381             0.724837   \n",
       "AGTGAACCATCCCGCT-8-s3d6        3183.0            4684             0.679547   \n",
       "\n",
       "                         blacklist_fraction  nucleosome_signal  \\\n",
       "TAGTTGTCACCCTCAC-1-s1d1            0.003473           0.642468   \n",
       "CTATGGCCATAACGGG-1-s1d1            0.003126           1.220679   \n",
       "CCGCACACAGGTTAAA-1-s1d1            0.001284           0.692573   \n",
       "TCATTTGGTAATGGAA-1-s1d1            0.000714           0.633838   \n",
       "ACCACATAGGTGTCCA-1-s1d1            0.003352           0.727660   \n",
       "...                                     ...                ...   \n",
       "TAGTAAGCAACTAGGG-8-s3d6            0.000553           0.866142   \n",
       "TGGTCCTTCGGCTAGC-8-s3d6            0.000810           0.964006   \n",
       "CGCTTGCGTTGTTGGA-8-s3d6            0.000358           1.215227   \n",
       "ACCCTCCCAGCCAGTT-8-s3d6            0.000000           0.714286   \n",
       "AGTGAACCATCCCGCT-8-s3d6            0.001257           0.946565   \n",
       "\n",
       "                                cell_type  pseudotime_order_ATAC batch  \n",
       "TAGTTGTCACCCTCAC-1-s1d1     Naive CD20+ B                    NaN  s1d1  \n",
       "CTATGGCCATAACGGG-1-s1d1        CD14+ Mono                    NaN  s1d1  \n",
       "CCGCACACAGGTTAAA-1-s1d1            CD8+ T                    NaN  s1d1  \n",
       "TCATTTGGTAATGGAA-1-s1d1            CD8+ T                    NaN  s1d1  \n",
       "ACCACATAGGTGTCCA-1-s1d1        CD16+ Mono                    NaN  s1d1  \n",
       "...                                   ...                    ...   ...  \n",
       "TAGTAAGCAACTAGGG-8-s3d6               HSC                0.23367  s3d6  \n",
       "TGGTCCTTCGGCTAGC-8-s3d6  CD4+ T activated                    NaN  s3d6  \n",
       "CGCTTGCGTTGTTGGA-8-s3d6               pDC                    NaN  s3d6  \n",
       "ACCCTCCCAGCCAGTT-8-s3d6            CD8+ T                    NaN  s3d6  \n",
       "AGTGAACCATCCCGCT-8-s3d6  CD4+ T activated                    NaN  s3d6  \n",
       "\n",
       "[22463 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This object also stores metadata for the cells as a dataframe\n",
    "\n",
    "atac.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deepsnap",
   "language": "python",
   "name": "deepsnap"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
